src/node/rpc/forwarder.h

```diff
@@ -17,133 +17,135 @@ namespace ccf
-
-    virtual bool forward_command(
-      enclave::RPCContext& rpc_ctx,
-      NodeId from,
-      NodeId to,
-      CallerId caller_id,
-      const std::vector<uint8_t>& data) = 0;
-  };
-
-  class Forwarder : public AbstractForwarder
+  template <typename ChannelProxy>
+  class Forwarder : public enclave::AbstractForwarder
   {
   private:
-    enclave::AbstractRPCResponder& rpcresponder;
-    std::shared_ptr<NodeToNode> n2n_channels;
-    std::shared_ptr<enclave::RpcMap> rpc_map;
+    std::shared_ptr<enclave::AbstractRPCResponder> rpcresponder;
+    std::shared_ptr<ChannelProxy> n2n_channels;
+    std::shared_ptr<enclave::RPCMap> rpc_map;
+    NodeId self;
+
+    using IsCallerCertForwarded = bool;
 
   public:
     Forwarder(
-      enclave::AbstractRPCResponder& rpcresponder,
-      std::shared_ptr<NodeToNode> n2n_channels) :
+      std::shared_ptr<enclave::AbstractRPCResponder> rpcresponder,
+      std::shared_ptr<ChannelProxy> n2n_channels,
+      std::shared_ptr<enclave::RPCMap> rpc_map_) :
       rpcresponder(rpcresponder),
-      n2n_channels(n2n_channels)
+      n2n_channels(n2n_channels),
+      rpc_map(rpc_map_)
     {}
 
-    void initialize(std::shared_ptr<enclave::RpcMap> rpc_map_)
+    void initialize(NodeId self_)
     {
-      rpc_map = rpc_map_;
+      self = self_;
     }
 
     bool forward_command(
       enclave::RPCContext& rpc_ctx,
       NodeId from,
       NodeId to,
       CallerId caller_id,
-      const std::vector<uint8_t>& data)
+      const std::vector<uint8_t>& data,
+      const std::vector<uint8_t>& caller_cert)
     {
-      std::vector<uint8_t> plain(
-        sizeof(caller_id) + sizeof(rpc_ctx.client_session_id) +
-        sizeof(rpc_ctx.actor) + data.size());
+      IsCallerCertForwarded include_caller = false;
+      size_t size = sizeof(caller_id) + sizeof(rpc_ctx.client_session_id) +
+        sizeof(rpc_ctx.actor) + sizeof(IsCallerCertForwarded) + data.size();
+      if (!caller_cert.empty())
+      {
+        size += sizeof(size_t) + caller_cert.size();
+        include_caller = true;
+      }
+
+      std::vector<uint8_t> plain(size);
       auto data_ = plain.data();
       auto size_ = plain.size();
       serialized::write(data_, size_, caller_id);
       serialized::write(data_, size_, rpc_ctx.client_session_id);
       serialized::write(data_, size_, rpc_ctx.actor);
+      serialized::write(data_, size_, include_caller);
+      if (include_caller)
+      {
+        serialized::write(data_, size_, caller_cert.size());
+        serialized::write(data_, size_, caller_cert.data(), caller_cert.size());
+      }
       serialized::write(data_, size_, data.data(), data.size());
 
       ForwardedHeader msg = {ForwardedMsg::forwarded_cmd, from};
 
       return n2n_channels->send_encrypted(to, plain, msg);
     }
 
-    std::optional<std::pair<enclave::RPCContext, std::vector<uint8_t>>>
+    std::optional<std::tuple<enclave::RPCContext, NodeId, std::vector<uint8_t>>>
```

Comment:
since we are returning more than 2 things maybe it's better to make a struct and have the returned elements be named?